cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/SqlStructuredTableRegistry.java

```diff
@@ -16,40 +16,181 @@
+      return GSON.fromJson(specString, StructuredTableSpecification.class);
+    });
   }
 
   public void removeSpecification(StructuredTableId tableId) {
-    specMap.remove(tableId);
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable registry = context.getTable(REGISTRY);
+      registry.delete(Collections.singleton(Fields.stringField(TABLE_NAME_FIELD, tableId.getName())));
+    });
   }
 
   public boolean isEmpty() {
-    return specMap.isEmpty();
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable registry = context.getTable(REGISTRY);
+      try (CloseableIterator<StructuredRow> it = registry.scan(Range.all(), Integer.MAX_VALUE)) {
+        return !it.hasNext();
+      }
+    });
+  }
+
+  private void createRegistryTable() throws IOException {
+    // Use a no-op registry to create PostgresSqlStructuredTableAdmin.
+    // During a table creation, except for registerSpecification, no other methods on the
+    // registry will be called by PostgresSqlStructuredTableAdmin.
+    StructuredTableRegistry noOpRegistry = new StructuredTableRegistry() {
+      final UnsupportedOperationException exception =
+        new UnsupportedOperationException("Not expected to be called during creation of registry!");
+
+      @Override
+      public void initialize() {
+        throw exception;
+      }
+
+      @Override
+      public void registerSpecification(StructuredTableSpecification specification) {
+        // Do nothing
+      }
+
+      @Nullable
+      @Override
+      public StructuredTableSpecification getSpecification(StructuredTableId tableId) {
+        throw exception;
+      }
+
+      @Override
+      public void removeSpecification(StructuredTableId tableId) {
+        throw exception;
+      }
+
+      @Override
+      public boolean isEmpty() {
+        throw exception;
+      }
+    };
+
+    try {
+      // Create the table if needed
+      PostgresSqlStructuredTableAdmin admin =
+        new PostgresSqlStructuredTableAdmin(noOpRegistry, dataSource);
+      if (!admin.tableExists(REGISTRY)) {
+        LOG.info("Creating SQL table {}", REGISTRY);
+        admin.create(SPEC);
+      }
+    } catch (TableAlreadyExistsException e) {
+      // Looks like the table was created concurrently by some other process
+      LOG.debug(String.format("Got exception when trying to create table %s", REGISTRY), e);
+    }
+  }
+
+  private TransactionRunner createTransactionRunner() {
+    final UnsupportedOperationException exception =
+      new UnsupportedOperationException("Unexpected DDL operation during registry usage!!");
+    // Create a spec admin that only returns the spec for the registry while creating SqlTransactionRunner
+    StructuredTableAdmin specAdmin =
+      new StructuredTableAdmin() {
+        @Override
+        public void create(StructuredTableSpecification spec) {
+          throw exception;
+        }
+
+        @Nullable
```

Comment:
Since this is returning SPEC, it can no longer be nullable

Reply:
Fixed